File: Renesys_Coding_Challenge/tooManyVPs.py
from sets import Set
from collections import defaultdict

vps=set([])
numEmployees=dict({})
ceo="1"
directEmployees=defaultdict(list)

def computeImportance(vp,boss):

        global numEmployees
        global directEmployees

        if(boss in directEmployees):
                numEmployees[vp]+=len(directEmployees[boss])
                for employees in directEmployees[boss]:
                        computeImportance(vp,employees)




#opening file
f=open("vp_ego.adjacency.txt",'r')

for line in f:
	bossEmployee=line.split()
#	print bossEmployee[0]	
	if bossEmployee[0] == ceo:
		vps.add(bossEmployee[1])
		numEmployees[bossEmployee[1]]=0
	else:
		directEmployees[bossEmployee[0]].append(bossEmployee[1])

for vp in vps:
	computeImportance(vp,vp)
	#print "VP " + repr(vp) + " has " + repr(numEmployees[vp]) + " employees"		

print "VP", "Number of Employees"	
for vp in sorted(numEmployees, key=numEmployees.get):
	print vp, numEmployees[vp]

File: Renesys_Coding_Challenge/tooManyVPs.pl
$fileName = "vp_ego.adjacency.txt";

open FILE, "<", $fileName or die "Could not find file\n";

$hashCounter = {};
$hashBoss    = {};

$counter = 0;
@vps     = ();

while (<FILE>) {

	$counter++;
	chomp $_;

	@categories = split( ' ', $_ );

	#print "Reading $_\n";

	$boss     = $categories[0];
	$employee = $categories[1];

	if ( $boss != 0 && $employee != 0 ) {

		if ( exists $hashBoss{$employee} ) {
			die "Employee already encountered\n";
		}

		else {

			if ( $boss == 1 ) {
				push( @vps, $employee );
			}

			$hashBoss{$employee} = $boss;
			if ( exists $hashCounter{$boss} ) {

				$hashCounter{$boss}++;

			}

			else {

				$hashCounter{$boss} = 1;
			}

		}

	}

}

foreach(@vps){
	print "Number of employees of vp $_ are: $hashCounter{$_}\n";
}

for my $key ( keys %hashCounter ) {
	
	$node=$key;

	if ( $hashBoss{$node} != 1 && $node != 1 )
	{

		while ( $hashBoss{$node} != 1 ) {
			$node = $hashBoss{$node};
		}

		$hashCounter{$node} += $hashCounter{$key};
	}
	
}

$min         = $counter;
$leastValued = -1;

foreach (@vps) {

	$key   = $_;
	$value = $hashCounter{$key};

	if ( $value < $min ) {
		$min         = $value;
		$leastValued = $key;
	}

}

print "Least important VP is: $leastValued with $min employees\n";
